### PR TITLE
RuntimeAssumptions: fix NULL used in arithmetic

### PR DESCRIPTION
**Pull request by Hello71**
ISO C11 section 7.20.1.4 only guarantees conversions to and from void pointer

not sure what's going on here exactly. _key doesn't seem to be read anywhere else, so shouldn't it be a bool?

### CODE REVIEW

**Comment by dsouzai** [comment]
Hey,

> not sure what's going on here exactly. _key doesn't seem to be read anywhere else, so shouldn't it be a bool?

`_key` is a member of the parent class `OMR::RuntimeAssumption` [1]. The reason it is set to NULL is because of a double free bug that was fixed about a year ago; you can read the details over at [2].

Your change looks reasonable to me; however, you'll need to sign the Eclipse ECA in order for the change to get accepted into the project [3].

[1] https://github.com/eclipse/omr/blob/master/compiler/runtime/OMRRuntimeAssumptions.hpp
[2] https://github.com/eclipse/openj9/pull/6932
[3] https://github.com/eclipse/openj9#how-do-i-contribute

**Comment by Hello71** [comment]
> > not sure what's going on here exactly. _key doesn't seem to be read anywhere else, so shouldn't it be a bool?
> 
> `_key` is a member of the parent class `OMR::RuntimeAssumption` [1]. The reason it is set to NULL is because of a double free bug that was fixed about a year ago; you can read the details over at [2].

Ah, I see. I am used to C, so forgot to check class access.

> Your change looks reasonable to me; however, you'll need to sign the Eclipse ECA in order for the change to get accepted into the project [3].

I have already signed the ECA, but it was not recognized because I committed using the web editor. I have now amended the commit.

**Comment by Hello71** [comment]
also, for future reference, the initial driver of this PR was to fix a gcc warning.

File: runtime/compiler/runtime/RuntimeAssumptions.cpp [hunk 1]
```diff
@@ -85,10 +85,10 @@ TR_PatchNOPedGuardSiteOnClassPreInitialize::hashCode(char *sig, uint32_t sigLen)
 void
 TR_PatchNOPedGuardSiteOnClassPreInitialize::reclaim()
    {
-   TR_ASSERT_FATAL(_key != NULL, "Attempt to reclaim an already freed _key");
+   TR_ASSERT_FATAL((void*)_key != NULL, "Attempt to reclaim an already freed _key");
```
Comment (dsouzai) [inline runtime/compiler/runtime/RuntimeAssumptions.cpp:88]:
```suggestion
   TR_ASSERT_FATAL(reinterpret_cast<void *>(_key) != NULL, "Attempt to reclaim an already freed _key");
```

File: runtime/compiler/runtime/RuntimeAssumptions.cpp [hunk 2]
```diff
@@ -85,10 +85,10 @@ TR_PatchNOPedGuardSiteOnClassPreInitialize::hashCode(char *sig, uint32_t sigLen)
...
 
    jitPersistentFree((void*)_key);
-   _key = 0;
+   _key = (uintptr_t)(void*)0;
```
Comment (dsouzai) [inline runtime/compiler/runtime/RuntimeAssumptions.cpp:91]:
```suggestion
   _key = 0;
```
`_key` is of type `uintptr_t` so a `0` is fine.

File: runtime/compiler/runtime/RuntimeAssumptions.cpp (hunk 1 above)
Comment (keithc-ca) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp:88]: I suggest that neither of those changes are appropriate. The field `_key` is declared with type `uintptr_t` (a scalar type): `0` is the right value for a freed key.

Comment (Hello71) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp:88]: The ISO C++ standard defers to the ISO C standard, which gives no guarantees on the behavior of (u)intptr_t except that pointers to void may be converted to and from it. It is not guaranteed that all-zero pointers retain the same value when converted to (u)intptr_t. Furthermore, it is not guaranteed that null pointers are represented as all zeroes in the first place. However, as a practical matter, current implementations directly copy the byte representation as an integer, allow integer operations on them, represent null pointers as all zeroes, and are unlikely to change any of those behaviors or make optimization decisions based on them.

In conclusion, the cast is more correct, but any of these three alternatives are likely to be equivalent in practice.

File: runtime/compiler/runtime/RuntimeAssumptions.cpp (hunk 2 above)
Comment (Hello71) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp:91]: See my other comment, but in brief, I believe this is not strictly correct according to the standard, similar to `void *p; memset(&p, 0, sizeof(p));` being an incorrect way to initialize a null pointer to void.

**Comment by Hello71** [comment]
actually, after a brief search (I don't have a C++ IDE), it seems to me that _key should in fact be a `void *`, and instead `TR_RuntimeAssumptionTable` should use `uintptr_t` and `size_t` where appropriate. for example, `TR_RuntimeAssumptionTable::getBucketPtr` seems like it should take a `size_t`, not a `uintptr_t`.

**Comment by dsouzai** [comment]
> actually, after a brief search (I don't have a C++ IDE), it seems to me that _key should in fact be a void *, and instead TR_RuntimeAssumptionTable should use uintptr_t and size_t where appropriate. for example, TR_RuntimeAssumptionTable::getBucketPtr seems like it should take a size_t, not a uintptr_t.

Yeah, I'm not sure why `uintptr_t` was chosen for `_key` rather than a `void *`; this is very much an area containing legacy code. Changing the code to use a `void *` will require a coordinated set of PRs in both OMR and OpenJ9, but I don't have any problem with introducing such clean up if you're up for it.

Regarding `size_t` for `hashIndex` in `getBucketPtr`, `TR_RuntimeAssumptionTable::hashCode` is
```
   static uintptr_t hashCode(uintptr_t key)
      {return (key >> 2) * 2654435761u; } // 2654435761u is the golden ratio of 2^32
```
Changing it to a `size_t` might change the hashcode depending on what the size of a `size_t` is on a given platform, which in turn might have consequences on performance. Therefore, a change like that would need to be vetted for performance. As it stands, I don't think there's a functional correctness issue since the hashcode is just generated based on the numerical value of the key (which may originate as a pointer).

File: runtime/compiler/runtime/RuntimeAssumptions.cpp (hunk 1 above)
Comment (keithc-ca) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp:88]: If `(0 != (uintptr_t)NULL)` then we'll have lots of broken code.
I still favour `0`, for consistency with other places that use `uintptr_t`.

Comment (Hello71) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp:88]: yeah, I think we can ignore `(0 != (uintptr_t)NULL)` for now. it would break a huge amount of existing code in other projects too, and AFAIK doesn't have any good optimization rationale. @dsouzai are you OK with `0 != _key` here?

Comment (dsouzai) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp:88]: Yeah no issues.

**Comment by dsouzai** [comment]
jenkins compile xlinux jdk11

**Comment by dsouzai** [comment]
jenkins test sanity xlinux jdk11

**Comment by Hello71** [comment]
> Our convention is to put an r-value on the left of `==` and `!=` to avoid some classes of coding errors, so it should be `0 != _key`, but I won't insist.

ah yes, I meant to put a comment that I did it this way to match the rest of the file.
